### ISSUE TITLE
Consideration on dynamic batch sizes

### ISSUE THREAD

**Issue by KennethEnevoldsen** [huggingface/sentence-transformers#3570]
Hi have been running quite a few evaluations recently on the MTEB benchmark (maintainer), where we both have large/small documents and large/small models. 

Typically, it was once the case that every small model could reasonably be run with a batch size like 32 without issues, as the model itself is fairly small and the document was never too long due to truncation. However, for more recent models (which are what are mostly validated), which allow for larger documents (e.g. 8k tokens), I find myself typically running a batch size of 1-2 even though only a few of the documents require a batch size this low. I wonder if there is a way (I might have missed it in the docs) or if we could add a way to have dynamic batch sizes based on document length or number of tokens.

This could be added in a backward-compatible manner using a keyword argument and maintaining the current default behaviour

As always, a big fan of the package,
Keep up the good work!

**Comment by tomaarsen** [comment]
Hello!

I agree that the default of 32 is becoming more "problematic" as models grow bigger. 
Sadly, I think computing a dynamic batch size would have to be done taking many factors into consideration, e.g. the size of the input, the number of model parameters, the inherent efficiency of the underlying model architecture, the memory available in the model. And even with all of these factors available, it's unlikely that I'd be able to design a proper algorithm that maximizes efficiency without hitting memory limits. 

I believe some training libraries do have automatic batch size computation during training, but those are (I believe) simply computed by retrying different multiples of 2 until there's an OOM. The overhead this brings is much more reasonable during training, I can't introduce something like this for inference.

In short, I don't think I would be able to do an implementation like this justice, so it might be better to leave it to the end user.

- Tom Aarsen

**Comment by KennethEnevoldsen** [comment]
Yeah, I can see the problem. For training, these trade-offs are much more reasonable as well since you make an effort to make consistent batch sizes, while this is typically not done during inference - though I support we could sort the documents based on lengths to reduce the influence of padding.

Often, in many cases, it is typically the case that 90% of batches are fine. Would it be possible to allow the user to specify their own batches:

```
texts = [["short", "short", "short"], ["very very long"]]
```

Or providing a dataloader?

**Comment by tomaarsen** [comment]
I'm a bit hesitant to build that functionality into the `encode` method. It's already such a "monolith" method with a ton of parameters/options. What you're proposing is certainly possible, though, by calling `model.tokenize` & `model.forward` instead of `model.encode`. This should give you the freedom to customize your batches.

- Tom Aarsen

**Comment by KennethEnevoldsen** [comment]
Yeah, tokenize and forward is probably a better way to go - I would just be afraid that I would miss model-specific adaptations (e.g. prompts). Where in the flow are they applied?

**Comment by tomaarsen** [comment]
You're right, you might miss some things. Prompts and Matryoshka truncation are included in `encode`. 

- Tom Aarsen

**Comment by KennethEnevoldsen** [comment]
Right, so both pre-processing and post-processing. I probably won't look deeper into this now (too high complexity compared to gain) so I will close this
